### PR TITLE
Feature/resource spawning

### PR DESCRIPTION
**Pull request by Hatzman91**
Features:
Resource Prefab
Example Resource Type

Resource Spawner Prefab
Resource Locations Editor
Spawning Based on ResourceType on Spawner

### CODE REVIEW

File: .vsconfig [hunk 1]
```diff
@@ -0,0 +1,6 @@
+﻿{
```
Comment (PadPalon) [inline .vsconfig:1]: Add VS specific files to .gitignore

File: Assets/Scripts/Player/Development/ResourceEditor.cs [hunk 2]
```diff
@@ -0,0 +1,137 @@
+using UnityEngine;
+
+#if UNITY_EDITOR
+using UnityEditor;
+#endif
+
+[CustomEditor(typeof(ResourceSpawner)), CanEditMultipleObjects]
+public class ResourceEditor : Editor {
+
+    private ResourceSpawner _resourceSpawner;
```
Comment (PadPalon) [inline Assets/Scripts/Player/Development/ResourceEditor.cs:10]: Don't like the underscore convention to mark private fields, they add nothing of value

File: Assets/Scripts/Player/Development/ResourceEditor.cs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    public void OnSceneGUI() {
+        if (_resourceSpawner.transform == Selection.activeTransform) {
+            DrawSpawnHandles();
+        }
+
+        EditorGUI.BeginChangeCheck();
+        var typeToSpawn = _resourceSpawner.TypeToSpawn;
+        if (EditorGUI.EndChangeCheck()) {            
+            Undo.RecordObject(_resourceSpawner, "Spawner");
+            var viewedModelFilter = ((MeshFilter)_resourceSpawner.GetComponent("MeshFilter"));
```
Comment (PadPalon) [inline Assets/Scripts/Player/Development/ResourceEditor.cs]: Outer braces unnecessary

File: Assets/Scripts/Player/Development/ResourceEditor.cs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        } else {
+            NewSpawnpointButton();
+            CleanupSpanwLocationsButton();
+        }
+    }
+
+    /// <summary>
+    /// Start the spawning of resources
+    /// </summary>
+    private void StartSpawningUI() {
```
Comment (PadPalon) [inline Assets/Scripts/Player/Development/ResourceEditor.cs:52]: This should probably be tied to the round system that is now in place

File: Assets/Scripts/Player/Development/ResourceEditor.cs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+            var newResourceLocation = Instantiate(
+                resourceLocationPrefab,
+                instantiationPosition,
+                instantiationTransform.rotation,
+                _resourceSpawner.transform
+            );
+
+            newResourceLocation.name = $"{type} {_resourceSpawner.ResourceLocations.Count + 1}";
+            Undo.RegisterCreatedObjectUndo(newResourceLocation, "Create new spanw location");
+            var newSpanwLocation = newResourceLocation.GetComponent<ResourceSpawnLocation>();
```
Comment (PadPalon) [inline Assets/Scripts/Player/Development/ResourceEditor.cs]: typ: spanw -> spawn

File: Assets/Scripts/Player/Development/ResourceEditor.cs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    /// <summary>
+    /// Cleanup deleted or "None" spawnpoints
+    /// </summary>
+    private void CleanUpSpawnlocations() {
+        _resourceSpawner.ResourceLocations.RemoveAll(location => location == null || ReferenceEquals(location, null));
+    }
+
+    //private void OnDestroy() {
```
Comment (PadPalon) [inline Assets/Scripts/Player/Development/ResourceEditor.cs]: remove commented code

File: Assets/Scripts/Player/Resources/Resource.cs [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+            _weight = type.Weight;
+            _resourceSpawnLocation = startingPosition;
+            transform.rotation = Quaternion.LookRotation(_resourceSpawnLocation.transform.position);
+            _initialized = true;
+        } else {
+            Debug.LogError("Do not try to initialize enemy twice");
+        }
+    }
+    // Update is called once per frame
+    void Update()
```
Comment (PadPalon) [inline Assets/Scripts/Player/Resources/Resource.cs]: remove empty functions

File: Assets/Scripts/Player/Resources/Resource.cs [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+    public void Initialize(ResourceType type, ResourceSpawnLocation startingPosition) {
+        if (!_initialized) {
+            _type = type;
+            _weight = type.Weight;
+            _resourceSpawnLocation = startingPosition;
+            transform.rotation = Quaternion.LookRotation(_resourceSpawnLocation.transform.position);
+            _initialized = true;
+        } else {
+            Debug.LogError("Do not try to initialize enemy twice");
```
Comment (PadPalon) [inline Assets/Scripts/Player/Resources/Resource.cs]: typo: not an enemy

File: Assets/Scripts/Player/Resources/Resource.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
+using UnityEngine;
+
+public class Resource : MonoBehaviour
+{
+    private bool _initialized = false;
```
Comment (PadPalon) [inline Assets/Scripts/Player/Resources/Resource.cs]: same

File: Assets/Scripts/Player/Resources/ResourceSpawnLocation.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
+using UnityEngine;
+
+public class ResourceSpawnLocation : MonoBehaviour
+{
+
+    public bool IsFree {
```
Comment (PadPalon) [inline Assets/Scripts/Player/Resources/ResourceSpawnLocation.cs]: Any difference to a public get; private set; field?

File: Assets/Scripts/Player/Resources/ResourceSpawnLocation.cs [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    private bool _isFree = true;
+
+    public void OnDrawGizmos() {
+        Gizmos.DrawWireSphere(transform.position, .5f);
+    }
+       
+    private void OnTriggerExit(Collider collider) {
+
```
Comment (PadPalon) [inline Assets/Scripts/Player/Resources/ResourceSpawnLocation.cs:10]: empty lines at function start but not between functions here and below

File: Assets/Scripts/Player/Resources/ResourceSpawner.cs [hunk 12]
```diff
@@ -0,0 +1,55 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class ResourceSpawner : MonoBehaviour {
+
+    [field: SerializeField]
+    public ResourceType TypeToSpawn { get; private set; }
+
+    [SerializeField]
+    private float _timeBetweenSpawns;
```
Comment (PadPalon) [inline Assets/Scripts/Player/Resources/ResourceSpawner.cs]: same

File: Assets/Scripts/Player/Resources/ResourceSpawner.cs [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+            _timeSinceLastSpawn = 0f;
+            _amountSpawned++;
+        } else {
+            _timeSinceLastSpawn += Time.deltaTime;
+        }
+    }
+
+    public void ResourceTaken() {
+        _amountSpawned -= 1;
+        _timeSinceLastSpawn = 0f;
```
Comment (PadPalon) [inline Assets/Scripts/Player/Resources/ResourceSpawner.cs]: Should taking a resource really reset time to spawn?

File: Assets/Scripts/Player/Resources/ResourceType.cs [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+public class ResourceType : ScriptableObject {
+    [field: SerializeField]
+    public float Weight { get; private set; }
+
+    [field: SerializeField]
+    public float SpawnCap { get; private set; }
+
+    [field: SerializeField]
+    public float SpanwCooldown { get; private set; }
+}
```
Comment (PadPalon) [inline Assets/Scripts/Player/Resources/ResourceType.cs]: no empty line at EOF

File: Packages/packages-lock.json [hunk 15]
```diff
@@ -58,7 +58,7 @@
       "url": "https://packages.unity.com"
     },
     "com.unity.services.core": {
-      "version": "1.4.2",
+      "version": "1.3.1",
```
Comment (PadPalon) [inline Packages/packages-lock.json]: why downgrade?

**Review by PadPalon — COMMENTED** [review]
Run Editorconfig to adhere to defined codestyle in all files automatically

in #10 I've implemented a generalized interactable system as well as the start of a generalized pickup system. Maybe integrate with that.

In-Engine review outstanding, internet to shoddy to download change to my laptops Unity-

File: Assets/Scripts/Player/Development/ResourceEditor.cs [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+
+            _resourceSpawner.ResourceLocations.Add(newSpanwLocation);
+            PrefabUtility.RecordPrefabInstancePropertyModifications(_resourceSpawner);
+        }
+    }
+
+    /// <summary>
+    /// Button to clean Spawnlocations
+    /// </summary>
+    private void CleanupSpanwLocationsButton() {
```
Comment (Parolo41) [inline Assets/Scripts/Player/Development/ResourceEditor.cs]: Typo (Spanw)

File: Assets/Scripts/Player/Development/ResourceEditor.cs [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+        if (GUILayout.Button($"Cleanup Spawnlocations")) {
+            CleanUpSpawnlocations();
+        }
+    }
+
+    /// <summary>
+    /// Cleanup deleted or "None" spawnpoints
+    /// </summary>
+    private void CleanUpSpawnlocations() {
```
Comment (Parolo41) [inline Assets/Scripts/Player/Development/ResourceEditor.cs]: Is SpawnLocations one word or two words? (Sometimes written as SpawnLocation and sometimes Spawnlocation)

File: Assets/Scripts/Player/Development/ResourceEditor.cs (hunk 3 above)
Comment (PadPalon) [reply Assets/Scripts/Player/Development/ResourceEditor.cs]: Also, bigbrick seems to not have been committed

File: Assets/Scripts/Player/Development/ResourceEditor.cs (hunk 6 above)
Comment (Guhl23) [inline Assets/Scripts/Player/Development/ResourceEditor.cs]: unused

File: Assets/Scripts/Player/Resources/Resource.cs [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+            _type = type;
+            _weight = type.Weight;
+            _resourceSpawnLocation = startingPosition;
+            transform.rotation = Quaternion.LookRotation(_resourceSpawnLocation.transform.position);
+            _initialized = true;
+        } else {
+            Debug.LogError("Do not try to initialize enemy twice");
+        }
+    }
+    // Update is called once per frame
```
Comment (Guhl23) [inline Assets/Scripts/Player/Resources/Resource.cs]: unused

**Review by PadPalon — COMMENTED** [review]
Can't comment on files, so some more here from the in-engine test:

- Trying to start spawning resources in Resources scene throws errors
- ResourceSpawner scene seems empty

File: Assets/Scripts/Player/Development/ResourceEditor.cs (hunk 3 above)
Comment (Hatzman91) [reply Assets/Scripts/Player/Development/ResourceEditor.cs]: the whole thing was just a PoC to switch out meshes, which I scrapped later, but forgot to delete that code

**Comment by PadPalon** [comment]
Had to add a missing prefab to make it work locally. Rest seems fine enough for a first version.

**Review by PadPalon — DISMISSED** [review]

**Review by Parolo41 — APPROVED** [review]

**Review by Guhl23 — APPROVED** [review]

**Pull request merged**